File: RScripts/ringworld_analysis.py
import numpy as np
import pandas as pd
import statsmodels.api as sm
from scipy import stats

def ringworld_analysis():
    """
    Python version of Ringworld_analysis.R
    Generates qualitative ringworld data regarding surface vs edge atmospheric losses
    from particle numbers recaptured and escaped.
    
    v1.0, Sep 16 2025 Siyona Agarwal
    """
    # Ringworld parameters
    worldWidth = np.array([
        1600000, 100000, 10000, 1000, 1000, 1000, 1000, 1000, 1000, 1000,
        200, 2000, 1000000, 200, 3957.98762536996
    ])  # km

    worldGrav = np.array([
        1, 1, 1, 1, 0.25, 0.5, 1, 1.5, 2, 3,
        1, 1, 0.00956122530571083/9.81, 1, 1
    ])  # g

    worldRad = np.array([
        149598000, 149598000, 149598000, 149598000, 149598000,
        149598000, 149598000, 149598000, 149598000, 149598000,
        1000, 1854977.7, 9.4e13, 1854977.7, 395798.762536996
    ])  # km

    # Particle data
    recaptured = np.array([
        953897, 94555, 94414, 93713, 30106, 37635, 46811, 49349, 49701, 49799,
        90559, 94174, 43002, 90855, 94362
    ])

    escaped = np.array([
        4, 16, 78, 851, 682, 560, 422, 327, 275, 201,
        4307, 392, 27, 4251, 209
    ])

    # Process data
    escape_frac_list = []
    escape_frac_std_list = []
    fudge_factor = 0

    print("Processing data...\n")
    for i in range(len(worldGrav)):
        # Create binary arrays (0 = recaptured, 1 = escaped)
        rec_p = np.zeros(int(recaptured[i]))
        esc_p = np.ones(int(escaped[i]))
        
        alldata = np.concatenate([rec_p, esc_p])
        
        # Calculate statistics
        avg_escape = np.mean(alldata)
        std_escape = np.std(alldata, ddof=1)  # Sample standard deviation
        std_avg_escape = std_escape / np.sqrt(len(alldata))
        
        # Store results
        escape_frac_list.append(avg_escape)
        escape_frac_std_list.append(std_avg_escape * (1 - fudge_factor))
        
        # Print results
        print(f"Trial {i+1}:")
        print(f"Expected escape chance: {avg_escape:.6f}")
        print(f"Expected escape chance std. dev.: {std_avg_escape:.6f}\n")

    # Convert to numpy arrays for easier manipulation
    escape_frac_list = np.array(escape_frac_list)
    escape_frac_std_list = np.array(escape_frac_std_list)

    # Prepare data for regression
    X = np.column_stack([
        np.log10(worldGrav),
        np.log10(worldWidth),
        np.log10(worldRad)
    ])
    X = sm.add_constant(X)  # Adds a constant term to the predictor

    # Fit linear models
    print("\nFitting linear models...\n")
    
    # Normal model
    y_normal = np.log10(escape_frac_list)
    model_normal = sm.OLS(y_normal, X).fit()
    
    # High estimate model
    y_high = np.log10(escape_frac_list + escape_frac_std_list)
    model_high = sm.OLS(y_high, X).fit()
    
    # Low estimate model
    y_low = np.log10(np.maximum(1e-10, escape_frac_list - escape_frac_std_list))  # Avoid log(0)
    model_low = sm.OLS(y_low, X).fit()

    # Print results
    print("=== Normal Model ===")
    print(model_normal.summary())
    print("\n=== High Estimate Model ===")
    print(model_high.summary())
    print("\n=== Low Estimate Model ===")
    print(model_low.summary())
    
    # Print interpretation
    print("\n" + "="*80)
    print("INTERPRETATION OF RESULTS:")
    print("="*80)
    print("""
1. The intercept coefficient represents log10 of leakage rate per particle for 1g ringworlds
   that are 1 km in width and 1 km in radius. Currently predictable with ~30% error.
   
2. The log10(worldGrav) coefficient shows the polynomial relation between leakage and gravity:
   - Leakage ~ gravity^(-0.62 ± 0.03)
   - Since atmospheric mass is inversely proportional to gravity:
     lifespan ~ gravity^(-0.38 ± 0.03)
   
3. The log10(worldWidth) coefficient shows a statistically significant linear correlation 
   between width and lifetime (which is proportional to the inverse of escape rate).
   
4. The log10(worldRad) coefficient shows no significant correlation with escape rates.
""")

if __name__ == "__main__":
    ringworld_analysis()

File: Unused_scripts/Ringworld_analysis.R
#### Heading ####
# Ringworld_analysis.R
#
# This file is used to generate qualitative ringworld data regarding surface vs
# edge atmospheric losses from particle numbers recaptured and escaped generated
# using the main Python codes for the simulation. It outputs the ringworld size
# expected to generate 100% losses and the loss rate associated with the face.
#
# V2.0.1, Eric Comstock, 16/07/2025
# V2.0, Eric Comstock, 12/07/2025
# V1.2.1, Eric Comstock, 09/06/2025
# V1.2, Eric Comstock, 09/06/2025
# V1.1, Eric Comstock, 02/06/2025
# V1.0, Eric Comstock, 02/06/2025

#### Data input ####

# Freitas ringworld is an outlier - figure out why this is

#First the width variation tests, then Edwin's gravity tests, then the general tests are added

# Ringworld test widths (km)
worldWidth  = c(1600000,100000,10000,1000,1000,1000,1000,1000,1000,1000,
                200, 2000, 1000000, 200, 3957.98762536996)

# Ringworld gravity (g)
worldGrav  = c(1,1,1,1,0.25, 0.5, 1, 1.5, 2, 3,
               1, 1, 0.00956122530571083/9.81, 1, 1)

# Ringworld radius (km)
worldRad = c(149598000,149598000,149598000,149598000,149598000,149598000,149598000,149598000,149598000,149598000,
             1000,1854977.7, 9.4e13, 1854977.7, 395798.762536996)

# Number of particles recaptured
recaptured             = c(953897 , 94555 , 94414 , 93713,
                           30106, 37635, 46811, 49349, 49701, 49799,
                           90559, 94174, 43002, 90855, 94362)
# Number of particles escaped
escaped                = c(4, 16, 78, 851,682, 560, 422, 327, 275, 201,
                           4307, 392, 27, 4251, 209)

#### Process data ####

escape_frac_list       = c()                         # Fraction of particles escaped
escape_frac_std_list   = c()                         # Std. dev. of the above fraction

fudge_factor = 0# Should be zero in the final code - prevents infinity from showing up

for (i in 1: length(worldGrav))                 # Data processing
{
  rec_p                = integer(recaptured[i]) # Vector of zeros with length recaptured[i]
  esc_p                = integer(escaped[i]) + 1# Vector of ones with length escaped[i]
  
  alldata              = c(rec_p, esc_p)        # Concatenated vector with a 1 for
                                                #each escape, and a 0 for each recapture
  
  avg_escape           = mean(alldata)          # Expected value of the binary escape data
  std_escape           = sqrt(var(alldata))     # Std. dev. of the binary escape data
  
  std_avg_escape       = std_escape / sqrt(length(alldata))# Std. dev. of the Expected 
                                                #value of the binary escape data
  
  # Printing results for clarity
  print(paste0('Trial ', as.character(i)))
  print(paste0('Expected escape chance: ', as.character(avg_escape)))
  print(paste0('Expected escape chance std. dev.: ', as.character(std_avg_escape)))
  print('')
  
  escape_frac_list     = c(escape_frac_list, avg_escape)
  escape_frac_std_list = c(escape_frac_std_list, std_avg_escape) * (1 - fudge_factor)
}

# Calculating linear regressions for average, +1s.d., and -1s.d. escape fraction data
l.normal               = lm(log10(escape_frac_list) ~ log10(worldGrav) + log10(worldWidth) + log10(worldRad))
l.high                 = lm(log10(escape_frac_list+escape_frac_std_list) ~ log10(worldGrav) + log10(worldWidth) + log10(worldRad))
l.low                  = lm(log10(escape_frac_list-escape_frac_std_list) ~ log10(worldGrav) + log10(worldWidth) + log10(worldRad))

# No plotting this time

#### Linear regression and final results ####

# Print linear regression statistics
l.normal
l.high
l.low

# (Intercept) coefficient shows log10 of leakage rate per particle for 1g ringworlds
#that are 1 km in width and 1 km in radius. Right now we can predict this with a
#30% error, an it is around 10 km

# log10(worldGrav) coefficient shows polynomial relation between leakage and world
#gravity. Leakage ~ gravity ^ -0.62 +- 0.03, so since atmo mass is inversely proportional
#to gravity, lifespan ~ gravity ^ -0.38 +- 0.03

# log10(worldWidth) coefficient shows a statistically linear correlation between
#width and lifetime (which is proportional to the inverse of escape rate)

# log10(worldRad) coefficient shows no significant correlation at all
